### ISSUE TITLE
kyber1024 fails AVX2 compile

### ISSUE THREAD

**Issue by mouse07410** [PQClean/PQClean#323]
MacOS. Repo branch `no-flags-for-keccak4x`.

All 89 tests via `cd test; python3 test_testvectors.py` passed.

However, attempt to build kyber1024 and kyber1024-90s failed:
```
$ make SCHEME=kyber1024 IMPLEMENTATION=avx2 -C test
cd ../crypto_kem/kyber1024/avx2 && /Applications/Xcode.app/Contents/Developer/usr/bin/make
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o kem.o kem.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o poly.o poly.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o polyvec.o polyvec.c
as -o fq.o fq.s
<instantiation>:1:23: error: invalid register name
vpmulhw         %ymm1,%ymm210,%ymm12
                      ^~~~~~~
fq.s:19:1: note: while in macro instantiation
red16           2 10
^
<instantiation>:4:24: error: invalid register name
vpsubw          %ymm12,%ymm210,%ymm210
                       ^~~~~~~
fq.s:19:1: note: while in macro instantiation
red16           2 10
^
<instantiation>:1:23: error: invalid register name
vpmulhw         %ymm1,%ymm311,%ymm12
                      ^~~~~~~
fq.s:20:1: note: while in macro instantiation
red16           3 11
^
<instantiation>:4:24: error: invalid register name
vpsubw          %ymm12,%ymm311,%ymm311
                       ^~~~~~~
fq.s:20:1: note: while in macro instantiation
red16           3 11
^
<instantiation>:1:23: error: invalid register name
vpmulhw         %ymm1,%ymm412,%ymm12
                      ^~~~~~~
fq.s:21:1: note: while in macro instantiation
red16           4 12
^
<instantiation>:4:24: error: invalid register name
vpsubw          %ymm12,%ymm412,%ymm412
                       ^~~~~~~
fq.s:21:1: note: while in macro instantiation
red16           4 12
^
<instantiation>:1:23: error: invalid register name
vpmulhw         %ymm1,%ymm513,%ymm12
                      ^~~~~~~
. . . . .
[ lots more of such messages ]
. . . . .
fq.s:99:1: note: while in macro instantiation
fqmulprecomp    1,2,9 12
^
<instantiation>:1:16: error: invalid register name
vpmullw         %ymm1,%ymm1013,%ymm12
                      ^~~~~~~~
fq.s:100:1: note: while in macro instantiation
fqmulprecomp    1,2,10 13
^
<instantiation>:2:16: error: invalid register name
vpmulhw         %ymm2,%ymm1013,%ymm1013
                      ^~~~~~~~
fq.s:100:1: note: while in macro instantiation
fqmulprecomp    1,2,10 13
^
<instantiation>:4:16: error: invalid register name
vpsubw          %ymm12,%ymm1013,%ymm1013
                       ^~~~~~~~
fq.s:100:1: note: while in macro instantiation
fqmulprecomp    1,2,10 13
^
make[1]: *** [fq.o] Error 1
make: *** [../crypto_kem/kyber1024/avx2/libkyber1024_avx2.a] Error 2
```

~~Perhaps, instead of `as` it should invoke `$(CC)` and pass CFLAGS???~~ *Sorry. Did not work.*

### Update

This was my workaround for `kyber1024`:
```diff
diff --git a/crypto_kem/kyber1024/avx2/Makefile b/crypto_kem/kyber1024/avx2/Makefile
index c5647c56..44b14c0d 100644
--- a/crypto_kem/kyber1024/avx2/Makefile
+++ b/crypto_kem/kyber1024/avx2/Makefile
@@ -1,5 +1,6 @@
 # This Makefile can be used with GNU Make or BSD Make
 
+AS=yasm -m amd64 -f macho64 -p gas 
 LIB=libkyber1024_avx2.a
 HEADERS=api.h params.h poly.h polyvec.h reduce.h fq.inc cbd.h consts.h ntt.h shuffle.inc verify.h indcpa.h rejsample.h symmetric.h fips202x4.h
 OBJECTS=kem.o poly.o polyvec.o fq.o shuffle.o cbd.o ntt.o invntt.o basemul.o consts.o \
```

Two questions:
1. ~~Would it be possible to tell the build/test setup what to use for `AS`, and what flags to pass to it?~~ *I keep forgetting that for `make` `X=xx` means I can set env var X and all would be well.*
2. How/why did the tests compile and run successfully? Or was kyber1024 not included in the tests?

Here's what I got trying some tests manually:
```
$ python test_functest.py -k "kyber1024 and clean"
====================================== test session starts =======================================
platform darwin -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/ur20980/src/PQClean/test, configfile: pytest.ini
collected 328 items / 324 deselected / 4 selected                                                

test_functest.py ..ss                                                                      [100%]

========================== 2 passed, 2 skipped, 324 deselected in 9.31s ==========================
Cleaning up testcases directory
$ python test_functest.py -k "kyber1024 and avx2"
====================================== test session starts =======================================
platform darwin -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/ur20980/src/PQClean/test, configfile: pytest.ini
collected 328 items / 328 deselected                                                             

==================================== 328 deselected in 0.04s =====================================
Cleaning up testcases directory
```

**Comment by mouse07410** [comment]
And more on kyber:
```
$ export AS='yasm -m amd64 -f macho64 -p gas '
$ make SCHEME=kyber1024-90s IMPLEMENTATION=clean
make: Nothing to be done for `all'.
$ make SCHEME=kyber1024-90s IMPLEMENTATION=avx2
cd ../crypto_kem/kyber1024-90s/avx2 && /Applications/Xcode.app/Contents/Developer/usr/bin/make
yasm -m amd64 -f macho64 -p gas  -o fq.o fq.s
yasm -m amd64 -f macho64 -p gas  -o shuffle.o shuffle.s
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o cbd.o cbd.c
yasm -m amd64 -f macho64 -p gas  -o ntt.o ntt.s
yasm -m amd64 -f macho64 -p gas  -o invntt.o invntt.s
yasm -m amd64 -f macho64 -p gas  -c -o basemul.o basemul.S
yasm: warning: unrecognized option `-c'
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o consts.o consts.c
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o verify.o verify.c
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o indcpa.o indcpa.c
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o rejsample.o rejsample.c
clang -mavx2 -maes -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o aes256ctr.o aes256ctr.c
ar -r libkyber1024-90s_avx2.a kem.o poly.o polyvec.o fq.o shuffle.o cbd.o ntt.o invntt.o basemul.o consts.o verify.o indcpa.o rejsample.o aes256ctr.o
ar: creating archive libkyber1024-90s_avx2.a
mkdir -p ../bin
clang -O3 -Wall -Wextra -Wpedantic -Wvla -Werror -std=c99 -Wundef -Wshadow -Wcast-align -Wpointer-arith -Wmissing-prototypes -fstrict-aliasing -fno-common -pipe -I../common  -DNTESTS=1 -DPQCLEAN_NAMESPACE=PQCLEAN_KYBER102490S_AVX2 -I../crypto_kem/kyber1024-90s/avx2 crypto_kem/functest.c ../common/aes.c ../common/sha2.c ../common/fips202.c ../common/nistseedexpander.c ../common/sp800-185.c ../common/randombytes.c -o ../bin/functest_kyber1024-90s_avx2 -L../crypto_kem/kyber1024-90s/avx2 -lkyber1024-90s_avx2
Undefined symbols for architecture x86_64:
  "_PQCLEAN_KYBER102490S_AVX2_basemul_acc_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_polyvec_pointwise_acc in libkyber1024-90s_avx2.a(polyvec.o)
  "_PQCLEAN_KYBER102490S_AVX2_basemul_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_basemul in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_csubq_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_compress in libkyber1024-90s_avx2.a(poly.o)
      _PQCLEAN_KYBER102490S_AVX2_poly_csubq in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_frommont_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_frommont in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_invntt_level6_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_invntt in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_invntt_levels0t5_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_invntt in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_ntt_level0_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_ntt in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_ntt_levels1t6_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_ntt in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_nttfrombytes_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_frombytes in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_ntttobytes_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_tobytes in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_nttunpack_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_nttunpack in libkyber1024-90s_avx2.a(poly.o)
  "_PQCLEAN_KYBER102490S_AVX2_reduce_avx", referenced from:
      _PQCLEAN_KYBER102490S_AVX2_poly_reduce in libkyber1024-90s_avx2.a(poly.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [../bin/functest_kyber1024-90s_avx2] Error 1
```

**Comment by thomwiggers** [comment]
Known issue: #251, will eventually fix that #307
